### DISCUSSION TITLE
classes/class_backbuffercopy

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#213] [User-contributed notes]
# classes/class_backbuffercopy

Inherits: Node2D< CanvasItem< Node< Object A node that copies a region of the screen to a buffer for access in shader code. Description: Node for back-buffering the currently-displayed screen. The ...

https://docs.godotengine.org/en/stable/classes/class_backbuffercopy.html

**Comment by bibizu** [comment]
It might be helpful to link this [tutorial](https://docs.godotengine.org/en/stable/tutorials/shaders/screen-reading_shaders.html), as it explains BackBuffer usage.

**Reply by Calinou** [reply]
Pull request opened:

- https://github.com/godotengine/godot/pull/98421

**Comment by hsunekichi** [comment]
A few clarifications for people struggling like me: 

During the drawing step, when the buffer is detected it forces a copy of the whole (or part) of the screen up to that point. Then, it offers it to the next shader that asks for a hint_screen_texture. You use this node to force a manual copy of the screen for a shader, forgetting about the automatic system that will only copy once the screen for the whole drawing step. 
- The shader receives a texture of the whole screen, even in RECT mode. To access it you always use SCREEN_UV, because the texture always covers the whole screen. The only difference in RECT mode is that only the portion of the rect will be filled, so you need to be sure that the SCREEN_UV doesn't index outside that region.
- In RECT mode, the rect uses viewport coordinates. To capture a sprite, you need to transform the sprite rect into viewport coordinates (pixels on the screen), and update the backBufferCopy with that each frame (because each frame the viewport transform will be different)
